module/apmprometheus/gatherer.go

```diff
@@ -80,8 +80,70 @@ func (g gatherer) GatherMetrics(ctx context.Context, out *apm.Metrics) error {
 					out.Add(name+".percentile."+strconv.Itoa(p), labels, q.GetValue())
 				}
 			}
+		case dto.MetricType_HISTOGRAM:
+			// For the bucket values, we follow the approach described by Prometheus's
+			// histogram_quantile function (https://prometheus.io/docs/prometheus/latest/querying/functions/#histogram_quantile)
+			// to achieve consistent percentile aggregation results:
+			//
+			// "The histogram_quantile() function interpolates quantile values by assuming a linear
+			// distribution within a bucket. (...) If a quantile is located in the highest bucket,
+			// the upper bound of the second highest bucket is returned. A lower limit of the lowest
+			// bucket is assumed to be 0 if the upper bound of that bucket is greater than 0. In that
+			// case, the usual linear interpolation is applied within that bucket. Otherwise, the upper
+			// bound of the lowest bucket is returned for quantiles located in the lowest bucket."
+			for _, m := range mf.GetMetric() {
+				h := m.GetHistogram()
+				// Total count for all values in this
+				// histogram. We want the per value count.
+				if h.GetSampleCount() == 0 {
+					continue
+				}
+				labels := makeLabels(m.GetLabel())
+				values := h.GetBucket()
+				midpoints := make([]float64, 0, len(values))
+				counts := make([]uint64, 0, len(values))
+				for i, b := range values {
+					count := b.GetCumulativeCount()
+					// we are excluding zero-count
+					// prometheus buckets.
+					if count == 0 {
+						continue
+					}
+					le := b.GetUpperBound()
+					if i == 0 {
+						if le > 0 {
+							le /= 2
+						}
+					} else if i == (len(values) - 1) {
```

Comment:
Based on my other comment, I suppose the final +Inf bucket is omitted if it's empty? In which case we'd need to check `math.IsInf(values[i].GetUpperBound(), 1)`

Reply:
I commented the code explaining the current situation as I found it while running test code. It appears the +Inf bucket isn't specifically included, but the histogram's total observations sum + count reflect values in the +Inf bucket, so it can be inferred.